lib/phoenix/router/helpers.ex

```diff
@@ -156,7 +156,17 @@ defmodule Phoenix.Router.Helpers do
       Generates url to a static asset given its file path.
       """
       def static_url(%Conn{private: private} = conn, path) do
-        static_url(private.phoenix_endpoint, path)
+        case private do
+          %{phoenix_static_url: %URI{} = uri} ->
+            %URI{uri | path: static_path(conn, path)}
```

Comment:
@chrismccord when I would like to introduce this change, would you rather have the return value of `static_path/2` override the path from the given override `URI` struct or would you prefer that the path in the struct is prepended to the return value of `static_path/2`?

Reply:
I think we just ignore the path in `put_router_url`, right? Anyway, I would make the behaviour the same. :)

Reply:
@josevalim I just gave this another thought and I think that discarding the path information when using an `%URI{}` struct might not be an optimal thing to do.

One example when this might not be optimal is when one would like to use an S3 bucket url in the form of `http(s)://s3.amazonaws.com/<bucket>/<object>` (or some other url with a path) to host the static files. While passing a string as an override to `put_static_url/2`, this string could of course also contain a path, but in order to have the same possibility available in case one passes a struct I guess appending the file path to the path inside the uri struct could be a sensible thing to do.

I could imagine that things are different for the routing mechanisms of phoenix though, but why is the path nillified within the `Phoenix.Router.Helpers.url/2` function when an `%URI{}` struct is passed when the endpoint itself can be configured to actually use a path?

Reply:
Good points. Let's keep the path as well then.

Reply:
Thank you for replying so quickly :-) I did and am currently revisiting the inline documentation.

I am still wondering if the `put_router_url` function behaves correctly when an URI struct is given? As seen [here](https://github.com/phoenixframework/phoenix/blob/master/test/phoenix/router/helpers_test.exs#L415), the test uses a path but it is not present within the final url being generated.